### PR TITLE
ENH: Increase `itk::SumOfSquaresImageFunction` class coverage

### PR DESCRIPTION
**Pull request by jhlegarreta**
- ENH: Increase `itk::SumOfSquaresImageFunction` class coverage
- STYLE: Prefer using `EXPECT_EQ` in GTest value check
- ENH: Check GTest test method return value

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

**Review by N-Dekker — CHANGES_REQUESTED** [review]
Requested changes. Please wait for details

File: Modules/Core/ImageFunction/test/itkSumOfSquaresImageFunctionGTest.cxx [hunk 1]
```diff
@@ -105,6 +106,11 @@ TestBasicObjectProperties()
   imageFunction->SetNeighborhoodRadius(radius);
   ITK_TEST_SET_GET_VALUE(radius, imageFunction->GetNeighborhoodRadius());
 
+  auto                                                                 size = TImage::SizeType::Filled(radius);
+  const itk::RectangularImageNeighborhoodShape<TImage::ImageDimension> shape(size);
+  unsigned int                                                         neighborhoodSize = shape.GetNumberOfOffsets();
+  ITK_TEST_SET_GET_VALUE(neighborhoodSize, imageFunction->GetNeighborhoodSize());
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply Modules/Core/ImageFunction/test/itkSumOfSquaresImageFunctionGTest.cxx:112]: I mimicked the expected value following the method calls:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/ImageFunction/include/itkSumOfSquaresImageFunction.h#L122
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkImageNeighborhoodOffsets.h#L44

The `radius` value here is 1 (and not 3), so although I understand what you mean, 1^N will not provide the expected value of 9.

**Comment by jhlegarreta** [comment]
> Requested change:
The call ITK_TEST_SET_GET_VALUE does not work well with GoogleTest. Instead, GoogleTest has its own macro's, like EXPECT_EQ, EXPECT_TRUE, EXPECT_THROW, etc. In this specific case, EXPECT_EQ seems most appropriate.

:+1:. Thanks changed in 0b48ada.

File: Modules/Core/ImageFunction/test/itkSumOfSquaresImageFunctionGTest.cxx (hunk 1 above)
[the remark this answers is not in the document]
Comment (N-Dekker) [reply Modules/Core/ImageFunction/test/itkSumOfSquaresImageFunctionGTest.cxx]: Ah, you're right. My bad! It should be (2*radius + 1) ^ 2.

**Review by N-Dekker — APPROVED** [review]
OK, thanks John, although I would suggest:

    EXPECT_EQ(imageFunction->GetNeighborhoodSize(), std::pow(2*radius + 1), TImage::ImageDimension)); 

The other ITK_EXERCISE_BASIC_OBJECT_METHODS call (in the same function) may also be removed, of course, and  `TestBasicObjectProperties()` may return `void`, instead of `int`. But I leave it up to you.

**Comment by jhlegarreta** [comment]
> OK, thanks John, although I would suggest:

> EXPECT_EQ(imageFunction->GetNeighborhoodSize(), std::pow(2*radius + 1), TImage::ImageDimension)); 

If all neighborhood sizes should have that value (or at least the rectangular shaped one), this
> std::pow(2*radius + 1, TImage::ImageDimension)

should be put into a method, so that it can be re-used (including the call stack that underlies in the method this patch set tests), and at that moment be used in any test.

As that is not the case (at least for the class at issue), I prefer to leave things as they are in the current commit.

> The other ITK_EXERCISE_BASIC_OBJECT_METHODS call (in the same function) may also be removed, of course,  and > TestBasicObjectProperties() may return void, instead of int. But I leave it up to you

Niels, the `TestBasicObjectProperties` method name is meaningful enough about what the method does. At the same time, `ITK_EXERCISE_BASIC_OBJECT_METHODS` is a fundamental call to exercise the basic methods of an object, ensures a better coverage, it has helped discover a non-negligible amount of bugs, and it helps better spotting classes that need their coverage to be increased among the +2900 files, 130K LoC . Whether there are better ways to quantitatively check the expected values is beyond the scope of this PR.

**Comment by N-Dekker** [comment]
Sorry Jon, I meant to say:

The other _ITK_TEST_SET_GET_VALUE_ call may also be removed.

The problem here is that ITK_TEST_SET_GET_VALUE indicates an error by doing `return EXIT_FAILURE`. This "return EXIT_FAILURE" is ignored by GoogleTest. A proper way to do such a check in a GoogleTest is by `EXPECT_EQ`.

**Comment by jhlegarreta** [comment]
> The problem here is that ITK_TEST_SET_GET_VALUE indicates an error by doing return EXIT_FAILURE. This "return EXIT_FAILURE" is ignored by GoogleTest. A proper way to do such a check in a GoogleTest is by EXPECT_EQ.

Changed in 89e53e5. `ITK_EXERCISE_BASIC_OBJECT_METHODS` still returns an `int` when failing: if the failure is ignored by GTest, or the test does not crash (in the worst of the cases), then we are being unable to identify underlying bugs when using GTest, which to me is a major drawback.

**Comment by N-Dekker** [comment]
> `ITK_EXERCISE_BASIC_OBJECT_METHODS` still returns an `int` when failing: if the failure is ignored by GTest, or the test does not crash (in the worst of the cases), then we are being unable to identify underlying bugs when using GTest, which to me is a major drawback.

As a quick-fix you could at least check the return value of those `TestBasicObjectProperties()` calls. Something like:

    EXPECT_EQ( TestBasicObjectProperties(), EXIT_SUCCESS );

GoogleTest would then certainly not ignore an EXIT_FAILURE from TestBasicObjectProperties() anymore. 

Just a suggestion.

**Pull request merged**